jax/_src/state/discharge.py

```diff
@@ -48,6 +53,81 @@ def discharge_state(jaxpr: core.Jaxpr, consts: Sequence[Any]
   new_jaxpr, _ , new_consts = pe.trace_to_jaxpr_dynamic(eval_jaxpr, in_avals)
   return new_jaxpr, new_consts
 
+@dataclasses.dataclass
+class Environment:
+  env: Dict[core.Var, Any]
+
+  def read(self, v: core.Atom) -> Any:
+    if type(v) is core.Literal:
+      return v.val
+    assert isinstance(v, core.Var)
+    return self.env[v]
+
+  def write(self, v: core.Var, val: Any) -> None:
+    self.env[v] = val
+
+class DischargeRule(Protocol):
+
+  def __call__(self, in_avals: Sequence[core.AbstractValue], *args: Any,
+      **params: Any) -> Tuple[Sequence[Optional[Any]], Sequence[Any]]:
+    ...
+
+_discharge_rules: dict[core.Primitive, DischargeRule] = {}
+
+def register_discharge_rule(prim: core.Primitive):
+  def register(f: DischargeRule):
+    _discharge_rules[prim] = f
+  return register
+
+def _has_refs(eqn: core.JaxprEqn):
+  return any(isinstance(v.aval, ShapedArrayRef) for v in eqn.invars)
+
+def _eval_jaxpr_discharge_state(jaxpr: core.Jaxpr, consts: Sequence[Any],
+                                *args: Any):
+  env = Environment({})
+
+  map(env.write, jaxpr.constvars, consts)
+  # Here some args may correspond to `Ref` avals but they'll be treated like
+  # regular values in this interpreter.
+  map(env.write, jaxpr.invars, args)
+
+  for eqn in jaxpr.eqns:
+    if _has_refs(eqn):
```

Comment:
Note to selves: Can we just check the eqn's effects tag here? Actually no! Consider a `for_loop` application which takes in a ref (eg by closing over it) but has an empty body. Because we say effects live on arrows, the empty body means there's no effect on the inner jaxpr or (given how we compute a `for_loop` application eqn's effects now) thus on the eqn. But we still want this function `_eval_jaxpr_discharge_state` to get rid of this ref!

In other words, we want this function to take a jaxpr like this:

```
{ lambda a:Ref . let
  in () } : (Ref -> ())
```

and produce a jaxpr which doesn't have any Ref-typed input binders.

So in a way we're really getting rid of Refs even more than we're discharging state.